test/integration/client-navigation/pages/nav/hash-changes-with-state.js

```diff
@@ -0,0 +1,64 @@
+import React, { Component } from 'react'
+import Router from 'next/router'
+
+let count = 0
+
+export default class SelfReload extends Component {
+  static getInitialProps({ res }) {
+    if (res) return { count: 0 }
+    count += 1
+
+    return { count }
+  }
+
+  handleAClick = () => {
+    Router.push(
+      '/nav/hash-changes-with-state',
+      '/nav/hash-changes-with-state#',
```

Comment:
Can we update this to actually change the hash to a new value to explain the use case a little more? 

The hash always being the same makes it seem like updating these values in `history.state` is the wrong place and could be done more so in component level state 

Are you trying to store something in `history.state` so that it's available when navigating back through history even from an external site? 

Overall seems to make sense to add the `options` to this `changeState` for consistency just trying to understand use case here better

Reply:
@ijjk yes, I'm trying to store something in `history.state` so that it's available when navigating forward and backward through history, but also because of how I've set up routing for a modal.

Background:
I've just finished refactoring my website over to NextJs from Create React App and React Router.  In React Router, you can pass the history state with every navigation change without restriction.  You can do the same in NextJS's router except for when the nav change is only a hash change, hence this PR.

My use case is too complicated for a test, but I'll describe it here so you can see why I need it.

I have an Instagram style modal where the modal has its own route.  And when you refresh the page, the component previously in a modal sitting above a list is now its own page.  

Modal version: https://prnt.sc/qpkqsg
Page version: https://prnt.sc/qpkr25

Because this is a modal, I'm already using `history.state` to store the `itemId` as follows:

```typescript
    Router.push(
      {
        pathname: '/[vendorName]/inventory',
        query: Router.query
      },
      {
        pathname: `/${item.vendor.name}/inventory/${item.id}`
      },
      {
        shallow: true,
        scroll: false,
        docId: item.id // must be docId to open modal and use modal carousel
        // using this docId to load the modal from the list component
      }
    );
```

Note this modal has tabs on the top right: Main, More, Advanced, Insights

When clicking in the "Price" box, a click function redirects the user to the More tab, which is on the same route but has its own hash `<href>#more`.

The Price box exists on the `#more` tab, which is where we want the user to edit the price (instead of the `#main` tab.  When this redirect takes place, we want to auto-focus the `#more` Price box.  See these two steps:

1. User clicks Price in `#main`  https://prnt.sc/qpkuum
2. Click function redirects to `#more` and auto-focuses in the new Price box: https://prnt.sc/qpkvev

The click function for the Price box on the `#main` form looks like this:

```typescript
      Router.push(
        window.history.state.url,
        window.location.pathname + '#more',
        {
          ...window.history.state.options, // keep the docId and shallow=true state
          autoFocus: 'price' // tell the `#more` form to auto-focus the price box on mount
        }
      );
```

See that `autoFocus: 'price'`?  That's history state I need.  We need it because:

1. We need to maintain the existing history state because of how the modal routing works.
2. We want to return to that auto-focused state when navigating back.  
3. It's simpler to pass this state through the history than to manage it from the parent component of all tabbed components.

Reply:
@ijjk I'm happy to update the test to make the use case more clear, but I'm unsure what you're looking for.  I imagine you want to keep your tests minimal and fast, so adding modals and tabbed components might be too complex.  But, happy to change it according to your recommendation.

Reply:
@ijjk Btw, to get this modal to work, I'm using the change you made in PR #9837

Reply:
Ah I see, yeah I don't think we need to make the test more verbose since your above explanation seems to describe the use case pretty well and the test does ensure the state is being passed and we won't regress on it without breaking that test 👍